### PR TITLE
fix(AAiT-backend-2B): Notification feature

### PR DESCRIPTION
**Pull request by petrosbeyene**
Add notification feature

### CODE REVIEW

File: aait/backend/group-2B/Application/DTOs/Notifications/INotificationDto.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+namespace SocialSync.Application.DTOs.Notifications;
+
+public enum NotificationType
```
Comment (abrahamshimekt) [inline aait/backend/group-2B/Application/DTOs/Notifications/INotificationDto.cs:3]: remove the enum

File: aait/backend/group-2B/Application/DTOs/Notifications/INotificationDto.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+{
+    Like,
+    Follow
+}
+public interface INotificationDto
+{
+    public int SenderId { get; set; }
+
+    public NotificationType NotificationType{get; set;}
```
Comment (abrahamshimekt) [inline aait/backend/group-2B/Application/DTOs/Notifications/INotificationDto.cs:12]: make the notification type string

File: aait/backend/group-2B/Application/DTOs/Notifications/NotificationCreateDto.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+namespace SocialSync.Application.DTOs.Notifications;
+
+public class NotificationCreateDto:INotificationDto
+{
+    public int SenderId { get; set; }
+
+    public NotificationType NotificationType{get; set;}
```
Comment (abrahamshimekt) [inline aait/backend/group-2B/Application/DTOs/Notifications/NotificationCreateDto.cs:7]: make the NotificationType string

File: aait/backend/group-2B/Application/DTOs/Notifications/NotificationDto.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+
+namespace SocialSync.Application.DTOs.Notifications;
+
+public class NotificationDto : BaseDto, INotificationDto
+{
+    public int SenderId { get; set; }
+    public int RecepientId { get; set; }
+
+    public NotificationType NotificationType{get; set;}
```
Comment (abrahamshimekt) [inline aait/backend/group-2B/Application/DTOs/Notifications/NotificationDto.cs:11]: make the type of NotificationType string

File: aait/backend/group-2B/Application/DTOs/Notifications/NotificationListDto.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+
+namespace SocialSync.Application.DTOs.Notifications;
+public class NotificationListDto : INotificationDto
+{
+    public int SenderId { get; set; }
+    public int RecepientId { get; set; }
+    public NotificationType NotificationType{get; set;}
```
Comment (abrahamshimekt) [inline aait/backend/group-2B/Application/DTOs/Notifications/NotificationListDto.cs:7]: make the type of NotificationType string

File: aait/backend/group-2B/Application/Features/Notifications/Handlers/Commands/CreateNotificationCommandHandler.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+        public CreateNotificationCommandHandler(IUnitOfWork unitOfWork, IMapper mapper)
+        {
+            _unitOfWork = unitOfWork;
+            _mapper = mapper;
+        }
+
+        public async Task<int> Handle(CreateNotificationCommand request, CancellationToken cancellationToken)
+        {
+            // add any necessary validation here before creating the notification
```
Comment (abrahamshimekt) [inline aait/backend/group-2B/Application/Features/Notifications/Handlers/Commands/CreateNotificationCommandHandler.cs:23]: add validation before creating

File: aait/backend/group-2B/Application/Features/Notifications/Handlers/Commands/CreateNotificationCommandHandler.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+            var notification = _mapper.Map<Notification>(request.NotificationCreateDto);
+            notification = await _unitOfWork.NotificationRepository.AddAsync(notification);
+
```
Comment (abrahamshimekt) [inline aait/backend/group-2B/Application/Features/Notifications/Handlers/Commands/CreateNotificationCommandHandler.cs:27]: check whether the database is affected or not

File: aait/backend/group-2B/Application/Features/Notifications/Handlers/Commands/DeleteNotificationCommandHandler.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            _unitOfWork = unitOfWork;
+        }
+
+        public async Task<Unit> Handle(DeleteNotificationCommand request, CancellationToken cancellationToken)
+        {
+            // add any necessary validation here before deleting the notification
+            Notification notification = await _unitOfWork.NotificationRepository.GetAsync(request.NotificationId);
+            await _unitOfWork.NotificationRepository.DeleteAsync(notification);
+
```
Comment (abrahamshimekt) [inline aait/backend/group-2B/Application/Features/Notifications/Handlers/Commands/DeleteNotificationCommandHandler.cs:22]: check whether the database is affected or not

File: aait/backend/group-2B/Application/Features/Notifications/Handlers/Queries/GetNotificationDetailRequestHandler.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        _unitOfWork = unitOfWork;
+        _mapper = mapper;
+    }
+
+    public async Task<NotificationDto> Handle(
+        GetNotificationDetailRequest request,
+        CancellationToken cancellationToken
+    )
+    {
+        var notification = await _unitOfWork.NotificationRepository.GetAsync(request.Id);
```
Comment (abrahamshimekt) [inline aait/backend/group-2B/Application/Features/Notifications/Handlers/Queries/GetNotificationDetailRequestHandler.cs:25]: check whether the notification is found or not

File: aait/backend/group-2B/Application/Features/Notifications/Handlers/Queries/GetNotificationListRequestHandler.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        _unitOfWork = unitOfWork;
+        _mapper = mapper;
+    }
+
+    public async Task<List<NotificationListDto>> Handle(
+        GetNotificationListRequest request,
+        CancellationToken cancellationToken
+    )
+    {
+        var notifications = await _unitOfWork.NotificationRepository.GetAll(request.UserId);
```
Comment (abrahamshimekt) [inline aait/backend/group-2B/Application/Features/Notifications/Handlers/Queries/GetNotificationListRequestHandler.cs:26]: check weather the notifications are null or empty

File: aait/backend/group-2B/Application/DTOs/Notifications/INotificationDto.cs (hunk 2 above)
Comment (petrosbeyene) [reply aait/backend/group-2B/Application/DTOs/Notifications/INotificationDto.cs]: Done

File: aait/backend/group-2B/Application/DTOs/Notifications/NotificationCreateDto.cs (hunk 3 above)
Comment (petrosbeyene) [reply aait/backend/group-2B/Application/DTOs/Notifications/NotificationCreateDto.cs]: Done

File: aait/backend/group-2B/Application/DTOs/Notifications/NotificationListDto.cs (hunk 5 above)
Comment (petrosbeyene) [reply aait/backend/group-2B/Application/DTOs/Notifications/NotificationListDto.cs]: Done

File: aait/backend/group-2B/Application/Features/Notifications/Handlers/Commands/CreateNotificationCommandHandler.cs (hunk 6 above)
Comment (petrosbeyene) [reply aait/backend/group-2B/Application/Features/Notifications/Handlers/Commands/CreateNotificationCommandHandler.cs:25]: Done

File: aait/backend/group-2B/Application/Features/Notifications/Handlers/Commands/DeleteNotificationCommandHandler.cs (hunk 8 above)
Comment (petrosbeyene) [reply aait/backend/group-2B/Application/Features/Notifications/Handlers/Commands/DeleteNotificationCommandHandler.cs:25]: Done

**Review by abrahamshimekt — APPROVED** [review]

**Pull request merged**